xdem/spstats.py

```diff
@@ -9,19 +9,92 @@
 import random
 import warnings
 from functools import partial
-from typing import Callable, Union
+from typing import Callable, Union, Optional
 
 import matplotlib.pyplot as plt
 import numpy as np
 import pandas as pd
 from scipy import integrate
 from scipy.optimize import curve_fit
+from skimage.draw import disk
+from scipy.interpolate import RegularGridInterpolator, LinearNDInterpolator, griddata
+from xdem.spatial_tools import nmad
 
 with warnings.catch_warnings():
     warnings.filterwarnings("ignore", category=DeprecationWarning)
     import skgstat as skg
     from skgstat import models
 
+def robust_interp_nd_binning(df: pd.DataFrame, var_names: list[str], statistic : Union[str, Callable, None] = nmad, min_count: Optional[int] = 100) -> Callable:
+    """
+    Estimate an interpolated function for an N-dimensional binning.
+    Relies on scipy.ndimage.map_coordinates: see for more information on input arguments.
+
+    :param df: dataframe of binned values according to explanatory variables
+    :param var_names: list of variable names to use
+    :param statistic: statistic to use
+    :param min_count: filter binned values with a sample count of less than this minimum count
+    :return:
+    """
+
+    # check that the dataframe contains what we need
+    for var in var_names:
+        if var not in df.columns:
+            raise ValueError('Variable "'+var+'" does not exist in the provided dataframe.')
+    statistic_name = statistic if isinstance(statistic,str) else statistic.__name__
+    if statistic_name not in df.columns:
+        raise ValueError('Statistic "' + statistic_name + '" does not exist in the provided dataframe.')
+    if min_count is not None and 'count' not in df.columns:
+        raise ValueError('Statistic "count" is not in the provided dataframe, necessary to use the min_count argument.')
+    if df.empty:
+        raise ValueError('Dataframe is empty.')
+
+    # keep only the binning in the number of dimensions corresponding to the length of the variables
+    df_sub = df[df.nd == len(var_names)]
+    if df_sub.empty:
+        raise ValueError('Dataframe does not contain the number of binned dimensions' + str(len(var_names))+' corresponding to the list of variables.')
+
+    # compute the middle values instead of bin interval
+    for var in var_names:
+        df_sub[var] = pd.IntervalIndex(df_sub[var]).mid.values
```

Comment:
Here, you assume that `df` has an `IntervalIndex`, right? What happens if it's not an `IntervalIndex`?

Reply:
I assume this is always used after `nd_binning`, which always returns an interval vector but not stored as index (because `MultiIndexes` are a bit of a nightmare to deal with...).

I ran into the problem that MultiIndexes are really hard to manipulate, and if I do:
```python
bins = (2,3,4,5)
df = pd.DataFrame()
df['variable'] = pd.IntervalIndex.from_breaks(bins)
```
then the interval is stored as a vector but you can't access it with the usual methods, for instance:
```python
df.variable.mid
```
does not work.
This is why I have to write:
```python
pd.IntervalIndex(df.slope).mid
```
which works.
If you have any ideas to simplify this!

Reply:
Hmm... Not sure I 100% understand the context, but what about just type checking the index? 

If it's an `IntervalIndex`, take the mid, and if not, just take the normal version? Or check if `hasattr(df.variable, "mid")`?

Reply:
Added with `isinstance(df_sub[var].dtype,pd.IntervalDtype)`